File: cluxmate/core/session_store.py
"""SessionStore — SQLite metadata + JSONL event-log persistence.

Session *metadata* (id, title, provider, model, cwd, group, pin, message_count)
stays in SQLite at ``<root>/cluxmate.db`` — the desktop and TUI share this schema.
Conversation *history* is now an append-only JSONL event log at
``<root>/sessions/<id>.jsonl`` (see :class:`~cluxmate.core.session_log_store.SessionLogStore`),
replacing the legacy ``<id>.json`` message-file format (D6: no migration).
"""

from __future__ import annotations

import os
import sqlite3
import time
import uuid
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from cluxmate.core.checkpoints import delete_shadow_repo_for_cwd
from cluxmate.core.session_log import SURFACE_EVENT_TYPES, SessionHeader, SessionLog
from cluxmate.core.session_log_store import (
    SessionLogStore,
    SessionNotFoundError,
    subagent_session_ids,
)


def _same_cwd(a: str, b: str) -> bool:
    """Whether two working directories are the same project.

    The shadow repo is keyed by the *resolved* absolute path, so sessions that
    stored the same directory with different spellings (relative vs absolute,
    trailing separator, symlink) must still count as one project. Mirror that
    normalization here rather than comparing raw strings.
    """
    if not a or not b:
        return False
    try:
        return Path(a).resolve() == Path(b).resolve()
    except OSError:
        return a == b


class SessionStore:
    """Manages session metadata in SQLite and history as a JSONL event log.

    ``root_dir`` defaults to ``~/.cluxmate``; pass an explicit path in tests so
    both the SQLite db and the ``sessions/`` directory live under one temp root.
    """

    def __init__(self, root_dir: str | Path | None = None):
        self._root = Path(root_dir) if root_dir is not None else Path.home() / ".cluxmate"
        self._session_dir = self._root / "sessions"
        self._session_dir.mkdir(parents=True, exist_ok=True)
        self._db_path = self._root / "cluxmate.db"
        self._log_store = SessionLogStore(self._session_dir)
        self._conn: sqlite3.Connection | None = None

    @property
    def log_store(self) -> SessionLogStore:
        """The JSONL event-log store (exposed so builders can persist subagent logs)."""
        return self._log_store

    # ── database connection ──────────────────────────────────────────

    @property
    def conn(self) -> sqlite3.Connection:
        if self._conn is None:
            self._conn = sqlite3.connect(str(self._db_path))
            self._conn.execute("PRAGMA journal_mode = WAL")
            self._conn.row_factory = sqlite3.Row
            self._init_schema()
        return self._conn

    def _init_schema(self):
        """Create tables at the final schema, then apply additive migrations."""
        self.conn.executescript("""
            CREATE TABLE IF NOT EXISTS sessions (
                id            TEXT PRIMARY KEY,
                title         TEXT NOT NULL DEFAULT 'New Session',
                provider      TEXT NOT NULL,
                model         TEXT NOT NULL,
                model_id      TEXT,
                api_type      TEXT,
                reasoning_effort TEXT,
                cwd           TEXT NOT NULL,
                created_at    TEXT NOT NULL,
                updated_at    TEXT NOT NULL,
                message_count INTEGER DEFAULT 0,
                group_id      TEXT REFERENCES groups(id) ON DELETE SET NULL,
                is_pinned     INTEGER NOT NULL DEFAULT 0
            );
            CREATE TABLE IF NOT EXISTS groups (
                id         TEXT PRIMARY KEY,
                name       TEXT NOT NULL,
                created_at TEXT NOT NULL,
                sort_order INTEGER NOT NULL DEFAULT 0,
                is_auto    INTEGER NOT NULL DEFAULT 0,
                path       TEXT
            );
        """)
        self.conn.commit()
        # v6: auto groups are keyed by their RESOLVED path, not name. The CREATE
        # TABLE above covers fresh DBs; ALTER here covers an existing table.
        cols = {row[1] for row in self.conn.execute("PRAGMA table_info(groups)").fetchall()}
        if "path" not in cols:
            self.conn.execute("ALTER TABLE groups ADD COLUMN path TEXT")
        # Backfill `path` for pre-existing auto groups from their first session's
        # cwd. Idempotent: only rows still lacking a path are touched.
        for row in self.conn.execute(
            "SELECT id FROM groups WHERE is_auto = 1 AND path IS NULL"
        ).fetchall():
            sess = self.conn.execute(
                "SELECT cwd FROM sessions WHERE group_id = ? ORDER BY updated_at ASC LIMIT 1",
                (row["id"],),
            ).fetchone()
            if sess and sess["cwd"]:
                resolved = os.path.realpath(sess["cwd"])
                self.conn.execute(
                    "UPDATE groups SET path = ? WHERE id = ?", (resolved, row["id"])
                )
        self.conn.commit()

    # ── group helpers (mirrors desktop session-store.ts) ────────────

    def _ensure_group_for_cwd(self, cwd: str) -> str | None:
        resolved = os.path.realpath(cwd) if cwd else ""
        name = os.path.basename(resolved)
        if not name:
            return None
        # Auto groups are keyed by their RESOLVED path (not name), so two
        # different directories that share a basename stay as separate projects.
        row = self.conn.execute(
            "SELECT id FROM groups WHERE is_auto = 1 AND path = ?", (resolved,)
        ).fetchone()
        if row:
            return row["id"]
        gid = uuid.uuid4().hex[:12]
        now = datetime.now(timezone.utc).isoformat()
        max_order = self.conn.execute(
            "SELECT COALESCE(MAX(sort_order), -1) AS mx FROM groups"
        ).fetchone()["mx"]
        self.conn.execute(
            "INSERT INTO groups (id, name, created_at, sort_order, is_auto, path) "
            "VALUES (?, ?, ?, ?, 1, ?)",
            (gid, name, now, max_order + 1, resolved),
        )
        self.conn.commit()
        return gid

    def _cleanup_auto_group(self, group_id: str | None):
        if not group_id:
            return
        row = self.conn.execute(
            "SELECT COUNT(*) AS cnt FROM sessions WHERE group_id = ?",
            (group_id,),
        ).fetchone()
        if row and row["cnt"] == 0:
            self.conn.execute(
                "DELETE FROM groups WHERE id = ? AND is_auto = 1",
                (group_id,),
            )
            self.conn.commit()

    # ── session CRUD ─────────────────────────────────────────────────

    def create(
        self,
        title: str,
        provider: str,
        model: str,
        working_dir: str,
        model_id: str = "",
        api_type: str = "",
        reasoning_effort: str | None = None,
    ) -> str:
        sid = uuid.uuid4().hex[:12]
        now = datetime.now(timezone.utc).isoformat()
        group_id = self._ensure_group_for_cwd(working_dir) if working_dir else None
        self.conn.execute(
            """INSERT INTO sessions
               (id, title, provider, model, model_id, api_type, reasoning_effort,
                cwd, created_at, updated_at, message_count, group_id)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 0, ?)""",
            (sid, title or "New Session", provider, model, model_id, api_type,
             reasoning_effort, working_dir, now, now, group_id),
        )
        self.conn.commit()
        # Materialize the JSONL event log header (D6: no legacy <id>.json).
        self._log_store.create(SessionHeader(
            id=sid,
            createdAt=int(time.time() * 1000),
            cwd=working_dir or None,
            provider=provider,
            model=model,
            apiType=api_type,
        ))
        return sid

    def load(self, session_id: str) -> dict[str, Any] | None:
        """Return session metadata + derived messages as a dict, or None."""
        row = self.conn.execute(
            "SELECT * FROM sessions WHERE id = ?", (session_id,)
        ).fetchone()
        if row is None:
            return None
        meta = dict(row)
        meta["is_pinned"] = bool(meta.get("is_pinned", 0))
        meta["working_dir"] = meta.pop("cwd", "")
        meta["messages"] = self.load_messages(session_id)
        return meta

    def load_log(self, session_id: str) -> SessionLog | None:
        """Reconstruct the live event log for a session (None if absent)."""
        try:
            header, events = self._log_store.load(session_id)
            return SessionLog.from_events(header, events)
        except SessionNotFoundError:
            return None

    def load_messages(self, session_id: str) -> list[dict]:
        """Derived provider-native messages (system excluded) from the event log."""
        log = self.load_log(session_id)
        return log.derive_messages() if log else []

    def append_events(self, session_id: str, events: list) -> None:
        """Durably append a contiguous batch of events and refresh metadata."""
        if not events:
            return
        self._log_store.append(session_id, events)
        # Increment message_count by the number of appended surface events.
        # (Recompute-from-log would be needed once surface `replace` compaction
        # events land — Phase 3b — since those shadow rather than add messages.)
        added = sum(1 for e in events if e.type in SURFACE_EVENT_TYPES)
        now = datetime.now(timezone.utc).isoformat()
        self.conn.execute(
            "UPDATE sessions SET message_count = message_count + ?, updated_at = ? "
            "WHERE id = ?",
            (added, now, session_id),
        )
        self.conn.commit()

    def sync_metadata(self, session_id: str) -> None:
        """Refresh message_count + updated_at from the JSONL log (metadata only).

        History is now persisted incrementally (IncrementalPersister), so the TUI
        calls this once at turn end to keep the SQLite side — session-list counts
        and recency ordering — in sync without re-append writes. Recomputing
        message_count from the surface is also correct under compaction (replace
        events shrink the surface rather than adding to it).
        """
        log = self.load_log(session_id)
        count = len(log.surface) if log else 0
        now = datetime.now(timezone.utc).isoformat()
        self.conn.execute(
            "UPDATE sessions SET message_count = ?, updated_at = ? WHERE id = ?",
            (count, now, session_id),
        )
        self.conn.commit()

    def set_title_if_default(self, session_id: str, title: str) -> None:
        """Set the title only when it is still the default placeholder."""
        row = self.conn.execute(
            "SELECT title FROM sessions WHERE id = ?", (session_id,)
        ).fetchone()
        if row and (not row["title"] or row["title"] == "New Session"):
            self.rename(session_id, title)

    def delete(self, session_id: str):
        # Record group + cwd before deleting so we can clean up auto-groups and
        # (when this was the last session in its directory) the shadow repo.
        row = self.conn.execute(
            "SELECT group_id, cwd FROM sessions WHERE id = ?", (session_id,)
        ).fetchone()
        group_id = row["group_id"] if row else None
        cwd = row["cwd"] if row else None
        self.conn.execute("DELETE FROM sessions WHERE id = ?", (session_id,))
        self.conn.commit()
        if group_id:
            self._cleanup_auto_group(group_id)
        # Cascade-delete subagent logs (each is its own <child>.jsonl reachable
        # from the parent's spawn events) so deleting a session doesn't leak them.
        for child_id in subagent_session_ids(self._log_store, session_id):
            self._log_store.delete(child_id)
        self._log_store.delete(session_id)
        # The shadow repo is per-directory and shared across sessions. Only once
        # no session in this directory remains are its checkpoint SHAs
        # unreferenced and the repo safe to remove; otherwise leave it in place
        # (other sessions' undo anchors depend on the unchanged history).
        if cwd:
            remaining = [
                r["cwd"]
                for r in self.conn.execute("SELECT cwd FROM sessions").fetchall()
            ]
            if not any(_same_cwd(cwd, other) for other in remaining):
                delete_shadow_repo_for_cwd(cwd)

    def list_all(self) -> list[dict[str, Any]]:
        """Return all sessions ordered by pinned first, then most-recently-updated."""
        rows = self.conn.execute(
            "SELECT * FROM sessions ORDER BY is_pinned DESC, updated_at DESC"
        ).fetchall()
        result = []
        for row in rows:
            d = dict(row)
            d["is_pinned"] = bool(d.get("is_pinned", 0))
            d["working_dir"] = d.pop("cwd", "")  # TUI code expects working_dir
            d.pop("group_id", None)  # TUI doesn't use groups yet
            result.append(d)
        return result

    def list_groups(self) -> list[dict[str, Any]]:
        rows = self.conn.execute(
            "SELECT * FROM groups ORDER BY sort_order ASC"
        ).fetchall()
        return [
            {**dict(r), "is_auto": bool(r["is_auto"])}
            for r in rows
        ]

    def update_cwd(self, session_id: str, cwd: str):
        old = self.conn.execute(
            "SELECT group_id FROM sessions WHERE id = ?", (session_id,)
        ).fetchone()
        old_group_id = old["group_id"] if old else None
        new_group_id = self._ensure_group_for_cwd(cwd) if cwd else None

        self.conn.execute(
            "UPDATE sessions SET cwd = ?, group_id = ?, updated_at = ? WHERE id = ?",
            (cwd, new_group_id, datetime.now(timezone.utc).isoformat(), session_id),
        )
        self.conn.commit()

        if old_group_id and old_group_id != new_group_id:
            self._cleanup_auto_group(old_group_id)

    def rename(self, session_id: str, title: str):
        self.conn.execute(
            "UPDATE sessions SET title = ?, updated_at = ? WHERE id = ?",
            (title, datetime.now(timezone.utc).isoformat(), session_id),
        )
        self.conn.commit()

    def set_reasoning_effort(self, session_id: str, effort: str | None):
        """Persist the session's selected reasoning level (None = provider default)."""
        self.conn.execute(
            "UPDATE sessions SET reasoning_effort = ?, updated_at = ? WHERE id = ?",
            (effort, datetime.now(timezone.utc).isoformat(), session_id),
        )
        self.conn.commit()

    def set_model(
        self, session_id: str, model_id: str, provider: str, model: str, api_type: str,
    ):
        """Persist the session's selected model (with denormalized display labels)."""
        self.conn.execute(
            "UPDATE sessions SET model_id = ?, provider = ?, model = ?, "
            "api_type = ?, updated_at = ? WHERE id = ?",
            (model_id, provider, model, api_type,
             datetime.now(timezone.utc).isoformat(), session_id),
        )
        self.conn.commit()

    def pin(self, session_id: str, pinned: bool):
        self.conn.execute(
            "UPDATE sessions SET is_pinned = ?, updated_at = ? WHERE id = ?",
            (1 if pinned else 0, datetime.now(timezone.utc).isoformat(), session_id),
        )
        self.conn.commit()

File: desktop/src/main/session-store.ts
import Database from 'better-sqlite3'
import { createHash } from 'crypto'
import { join, basename, dirname } from 'path'
import { app } from 'electron'
import { unlinkSync, existsSync, mkdirSync, copyFileSync, writeFileSync, readFileSync, rmSync, realpathSync } from 'fs'
import type { SessionMeta, CreateSessionParams, GroupMeta } from '../shared/types'

let _db: Database.Database | null = null

function dbPath(): string {
  return join(app.getPath('home'), '.cluxmate', 'cluxmate.db')
}

function migrateDbIfNeeded(): void {
  const newPath = dbPath()
  if (existsSync(newPath)) return

  // The db has lived in two different locations historically:
  // 1. %APPDATA%/cluxmate-desktop/cluxmate.db  (original electron-builder name)
  // 2. %APPDATA%/cluxmate/cluxmate.db         (current app.getPath('userData'))
  // Check both so users upgrading from any old version keep their sessions.
  const candidates = [
    join(dirname(app.getPath('userData')), 'cluxmate-desktop', 'cluxmate.db'),
    join(app.getPath('userData'), 'cluxmate.db'),
  ]
  for (const oldPath of candidates) {
    if (existsSync(oldPath)) {
      mkdirSync(dirname(newPath), { recursive: true })
      copyFileSync(oldPath, newPath)
      return
    }
  }
}

function getDb(): Database.Database {
  if (!_db) {
    migrateDbIfNeeded()
    const newPath = dbPath()
    mkdirSync(dirname(newPath), { recursive: true })
    _db = new Database(newPath)
    _db.pragma('journal_mode = WAL')
    _db.exec(`
      CREATE TABLE IF NOT EXISTS sessions (
        id            TEXT PRIMARY KEY,
        title         TEXT NOT NULL DEFAULT 'New Session',
        provider      TEXT NOT NULL,
        model         TEXT NOT NULL,
        cwd           TEXT NOT NULL,
        created_at    TEXT NOT NULL,
        updated_at    TEXT NOT NULL,
        message_count INTEGER DEFAULT 0
      )
    `)
    // v2 schema: pin the config model entry id (+ api_type snapshot) per
    // session. Older DBs predate these columns — add them if missing. Existing
    // rows get NULL, and ensureBridge falls back to the active model for those.
    const cols = new Set(
      (_db.prepare('PRAGMA table_info(sessions)').all() as { name: string }[])
        .map((c) => c.name)
    )
    if (!cols.has('model_id')) _db.exec('ALTER TABLE sessions ADD COLUMN model_id TEXT')
    if (!cols.has('api_type')) _db.exec('ALTER TABLE sessions ADD COLUMN api_type TEXT')
    // Per-session reasoning-level selection (null → provider default).
    if (!cols.has('reasoning_effort')) _db.exec('ALTER TABLE sessions ADD COLUMN reasoning_effort TEXT')

    // v3 schema: groups table + group_id on sessions.
    // Migrate from the earlier "projects" naming if present.
    const existingTables = new Set(
      (_db.prepare("SELECT name FROM sqlite_master WHERE type='table'").all() as { name: string }[])
        .map((r) => r.name)
    )
    if (existingTables.has('projects') && !existingTables.has('groups')) {
      _db.exec('ALTER TABLE projects RENAME TO groups')
      existingTables.add('groups')
    }
    if (!existingTables.has('groups')) {
      _db.exec(`
        CREATE TABLE IF NOT EXISTS groups (
          id         TEXT PRIMARY KEY,
          name       TEXT NOT NULL,
          created_at TEXT NOT NULL,
          sort_order INTEGER NOT NULL DEFAULT 0
        )
      `)
    }
    const pcols = new Set(
      (_db.prepare('PRAGMA table_info(sessions)').all() as { name: string }[])
        .map((c) => c.name)
    )
    if (pcols.has('project_id') && !pcols.has('group_id')) {
      _db.exec('ALTER TABLE sessions RENAME COLUMN project_id TO group_id')
    }
    if (!pcols.has('group_id') && !pcols.has('project_id')) {
      _db.exec('ALTER TABLE sessions ADD COLUMN group_id TEXT REFERENCES groups(id) ON DELETE SET NULL')
    }
    // v4: mark auto-created groups so they can be cleaned up when empty
    const gcols = new Set(
      (_db.prepare('PRAGMA table_info(groups)').all() as { name: string }[])
        .map((c) => c.name)
    )
    if (!gcols.has('is_auto')) {
      _db.exec('ALTER TABLE groups ADD COLUMN is_auto INTEGER NOT NULL DEFAULT 0')
    }
    // v6: auto groups are keyed by their RESOLVED working directory (`path`),
    // not by name — two different directories that share a basename must not
    // collapse into one project. `name` stays as the display label.
    if (!gcols.has('path')) {
      _db.exec('ALTER TABLE groups ADD COLUMN path TEXT')
    }
    // Backfill `path` for pre-existing auto groups from their first session's
    // cwd. Idempotent: only rows still lacking a path are touched.
    {
      const stale = _db.prepare(
        'SELECT id FROM groups WHERE is_auto = 1 AND path IS NULL'
      ).all() as { id: string }[]
      for (const g of stale) {
        const sess = _db.prepare(
          'SELECT cwd FROM sessions WHERE group_id = ? ORDER BY updated_at ASC LIMIT 1'
        ).get(g.id) as { cwd: string } | undefined
        if (sess?.cwd) {
          let resolved = sess.cwd
          try { resolved = realpathSync(sess.cwd) } catch { /* keep raw */ }
          _db.prepare('UPDATE groups SET path = ? WHERE id = ?').run(resolved, g.id)
        }
      }
    }
    // v5: pin sessions to the top within their group
    if (!pcols.has('is_pinned')) {
      _db.exec('ALTER TABLE sessions ADD COLUMN is_pinned INTEGER NOT NULL DEFAULT 0')
    }
  }
  return _db
}

function resolveCwd(cwd: string): string {
  try { return realpathSync(cwd) } catch { return cwd }
}

function _cwdGroupName(cwd: string): string | null {
  const base = basename(cwd)
  // Root paths yield empty basename — skip auto-grouping
  return base || null
}

function _ensureGroupForCwd(cwd: string): string | null {
  const resolved = resolveCwd(cwd)
  const name = _cwdGroupName(resolved)
  if (!name) return null
  const db = getDb()
  // Auto groups are keyed by their RESOLVED path (not name), so two different
  // directories that share a basename stay as separate projects.
  const existing = db.prepare(
    'SELECT id FROM groups WHERE is_auto = 1 AND path = ?'
  ).get(resolved) as { id: string } | undefined
  if (existing) return existing.id
  const id = crypto.randomUUID().replace(/-/g, '').slice(0, 12)
  const now = new Date().toISOString()
  const maxOrder = (db.prepare('SELECT COALESCE(MAX(sort_order), -1) FROM groups').get() as Record<string, number>)
  const order = Object.values(maxOrder)[0] + 1
  db.prepare(
    'INSERT INTO groups (id, name, created_at, sort_order, is_auto, path) VALUES (?, ?, ?, ?, 1, ?)'
  ).run(id, name, now, order, resolved)
  return id
}

function _cleanupAutoGroup(groupId: string | null) {
  if (!groupId) return
  const db = getDb()
  const cnt = (db.prepare('SELECT COUNT(*) as cnt FROM sessions WHERE group_id = ?').get(groupId) as { cnt: number }).cnt
  if (cnt === 0) {
    db.prepare('DELETE FROM groups WHERE id = ? AND is_auto = 1').run(groupId)
  }
}

// Sessions can store the same directory with different spellings (relative vs
// absolute, trailing separator, symlink). The shadow repo is keyed by the
// *resolved* path, so compare resolved paths rather than raw strings — matching
// the Python side's `_same_cwd`.
function sameCwd(a: string, b: string): boolean {
  if (!a || !b) return false
  try {
    return realpathSync(a) === realpathSync(b)
  } catch {
    return a === b
  }
}

// Delete the shadow-git checkpoint repo for a working directory, if one exists.
// The repo is keyed by sha1(resolve(cwd)) — the exact derivation the Python side
// uses (CheckpointManager) — and is shared across every session in that
// directory. Only call this once no session in the directory remains, or a still
// live session's stored checkpoint SHAs would dangle. A path that can no longer
// be resolved (realpathSync throws) is a safe no-op: the key can't be derived,
// and the 30-day retention sweep will reap any orphaned repo.
function deleteShadowRepoForCwd(cwd: string): void {
  try {
    const resolved = realpathSync(cwd)
    const digest = createHash('sha1').update(resolved).digest('hex')
    rmSync(join(app.getPath('home'), '.cluxmate', 'checkpoints', `${digest}.git`), {
      recursive: true, force: true, maxRetries: 3, retryDelay: 100,
    })
  } catch { /* directory no longer resolvable — leave the repo for retention */ }
}

export function listSessions(): SessionMeta[] {
  const db = getDb()
  return db.prepare(
    'SELECT * FROM sessions ORDER BY is_pinned DESC, updated_at DESC'
  ).all() as SessionMeta[]
}

export function createSession(params: CreateSessionParams): SessionMeta {
  const db = getDb()
  const id = crypto.randomUUID().replace(/-/g, '').slice(0, 12)
  const now = new Date().toISOString()
  const groupId = _ensureGroupForCwd(params.cwd)

  db.prepare(`
    INSERT INTO sessions (id, title, provider, model, model_id, api_type, cwd, created_at, updated_at, message_count, group_id)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 0, ?)
  `).run(id, params.title || 'New Session', params.provider, params.model, params.modelId, params.apiType, params.cwd, now, now, groupId)

  // Seed the JSONL event-log header so the Python agent's _load_or_create_log
  // finds an existing session to append to (rather than creating a fresh one).
  // `modelId` is the config entry id (SQLite-only); the header carries the
  // display provider/model/apiType, matching Python's SessionStore.create().
  const sessionsDir = join(app.getPath('home'), '.cluxmate', 'sessions')
  mkdirSync(sessionsDir, { recursive: true })
  writeFileSync(
    join(sessionsDir, `${id}.jsonl`),
    JSON.stringify({
      type: 'session', version: 0, id,
      createdAt: Date.now(),
      cwd: params.cwd,
      provider: params.provider,
      model: params.model,
      apiType: params.apiType,
    }) + '\n',
    'utf-8',
  )

  return db.prepare('SELECT * FROM sessions WHERE id = ?').get(id) as SessionMeta
}

export function getSession(id: string): SessionMeta | undefined {
  return getDb().prepare('SELECT * FROM sessions WHERE id = ?').get(id) as SessionMeta | undefined
}

export function updateSession(
  id: string,
  updates: Partial<Pick<SessionMeta, 'title' | 'message_count' | 'model_id' | 'api_type' | 'model' | 'provider' | 'reasoning_effort'>>
) {
  const sets: string[] = []
  const vals: unknown[] = []

  if (updates.title !== undefined) { sets.push('title = ?'); vals.push(updates.title) }
  if (updates.message_count !== undefined) { sets.push('message_count = ?'); vals.push(updates.message_count) }
  if (updates.model_id !== undefined) { sets.push('model_id = ?'); vals.push(updates.model_id) }
  if (updates.api_type !== undefined) { sets.push('api_type = ?'); vals.push(updates.api_type) }
  if (updates.model !== undefined) { sets.push('model = ?'); vals.push(updates.model) }
  if (updates.provider !== undefined) { sets.push('provider = ?'); vals.push(updates.provider) }
  if (updates.reasoning_effort !== undefined) { sets.push('reasoning_effort = ?'); vals.push(updates.reasoning_effort) }

  if (sets.length === 0) return

  sets.push("updated_at = ?")
  vals.push(new Date().toISOString())
  vals.push(id)

  getDb().prepare(`UPDATE sessions SET ${sets.join(', ')} WHERE id = ?`).run(...vals)
}

export function updateSessionCwd(id: string, cwd: string) {
  const db = getDb()
  const old = db.prepare('SELECT group_id FROM sessions WHERE id = ?').get(id) as { group_id: string | null } | undefined
  const oldGroupId = old?.group_id ?? null
  const newGroupId = _ensureGroupForCwd(cwd)

  db.prepare(
    'UPDATE sessions SET cwd = ?, group_id = ?, updated_at = ? WHERE id = ?'
  ).run(cwd, newGroupId, new Date().toISOString(), id)

  if (oldGroupId && oldGroupId !== newGroupId) {
    _cleanupAutoGroup(oldGroupId)
  }
}

export function renameSession(id: string, title: string) {
  getDb().prepare(
    'UPDATE sessions SET title = ?, updated_at = ? WHERE id = ?'
  ).run(title, new Date().toISOString(), id)
}

export function createGroup(name: string): GroupMeta {
  const db = getDb()
  const id = crypto.randomUUID().replace(/-/g, '').slice(0, 12)
  const now = new Date().toISOString()
  const maxOrder = (db.prepare('SELECT COALESCE(MAX(sort_order), -1) FROM groups').get() as { [key: string]: number })
  const order = maxOrder ? Object.values(maxOrder)[0] + 1 : 0
  db.prepare(
    'INSERT INTO groups (id, name, created_at, sort_order) VALUES (?, ?, ?, ?)'
  ).run(id, name, now, order)
  return db.prepare('SELECT * FROM groups WHERE id = ?').get(id) as GroupMeta
}

export function deleteGroup(id: string) {
  const db = getDb()
  db.prepare('UPDATE sessions SET group_id = NULL WHERE group_id = ?').run(id)
  db.prepare('DELETE FROM groups WHERE id = ?').run(id)
}

export function listGroups(): GroupMeta[] {
  return getDb().prepare('SELECT * FROM groups ORDER BY sort_order ASC').all() as GroupMeta[]
}

export function renameGroup(id: string, name: string) {
  getDb().prepare('UPDATE groups SET name = ? WHERE id = ?').run(name, id)
}

export function getGroup(id: string): GroupMeta | undefined {
  return getDb().prepare('SELECT * FROM groups WHERE id = ?').get(id) as GroupMeta | undefined
}

export function moveSessionToGroup(sessionId: string, groupId: string | null) {
  const db = getDb()
  // Remember source group before moving so we can clean up if it's now empty
  const session = db.prepare('SELECT group_id FROM sessions WHERE id = ?').get(sessionId) as { group_id: string | null } | undefined
  const oldGroupId = session?.group_id ?? null

  if (groupId === null) {
    db.prepare('UPDATE sessions SET group_id = NULL, updated_at = ? WHERE id = ?')
      .run(new Date().toISOString(), sessionId)
  } else {
    db.prepare('UPDATE sessions SET group_id = ?, updated_at = ? WHERE id = ?')
      .run(groupId, new Date().toISOString(), sessionId)
  }

  // If the session left an auto-group and that group is now empty, remove it
  if (oldGroupId && oldGroupId !== groupId) {
    _cleanupAutoGroup(oldGroupId)
  }
}

// Move a session out of a user-created group back to the auto group for its
// working directory (its "project"). Recreates the auto group if it was cleaned
// up when the session left (e.g. it was the project's last session).
export function moveSessionToProject(sessionId: string) {
  const db = getDb()
  const session = db.prepare('SELECT cwd FROM sessions WHERE id = ?').get(sessionId) as { cwd: string } | undefined
  if (!session) return
  moveSessionToGroup(sessionId, _ensureGroupForCwd(session.cwd))
}

export function pinSession(id: string, pinned: boolean) {
  getDb().prepare('UPDATE sessions SET is_pinned = ?, updated_at = ? WHERE id = ?')
    .run(pinned ? 1 : 0, new Date().toISOString(), id)
}

export function deleteSession(id: string) {
  const db = getDb()
  const session = db.prepare('SELECT group_id, cwd FROM sessions WHERE id = ?').get(id) as { group_id: string | null, cwd: string } | undefined
  db.prepare('DELETE FROM sessions WHERE id = ?').run(id)
  // Clean up auto-group if this was its last session
  if (session) _cleanupAutoGroup(session.group_id)
  const base = join(app.getPath('home'), '.cluxmate', 'sessions')
  // Delete subagent logs first (each is its own <child>.jsonl reachable from the
  // parent's spawn events), then the parent JSONL + display transcript.
  for (const childId of subagentSessionIds(id, base)) {
    try { unlinkSync(join(base, `${childId}.jsonl`)) } catch { /* may not exist */ }
  }
  for (const f of [`${id}.jsonl`, `${id}.display.json`]) {
    try { unlinkSync(join(base, f)) } catch { /* may not exist */ }
  }
  // The shadow repo is per-directory and shared across sessions. Only once no
  // session in this directory remains are its checkpoint SHAs unreferenced and
  // the repo safe to remove; otherwise leave it (other sessions' undo anchors
  // depend on the unchanged history).
  if (session?.cwd) {
    const remaining = db.prepare('SELECT cwd FROM sessions').all() as { cwd: string }[]
    if (!remaining.some((r) => sameCwd(session!.cwd, r.cwd))) {
      deleteShadowRepoForCwd(session.cwd)
    }
  }
}

// Collect every subagent session id reachable from a session's append-only JSONL
// log, so deleting a session cascades to its subagent logs. Reads + parses the
// JSONL directly (no Python process required): each line is an event envelope
// `{seq, time, type, data, ...}`; a `subagent/spawn` event's `data.session_id`
// is the child log's filename under the same sessions dir.
function subagentSessionIds(id: string, base: string, seen: Set<string> = new Set()): string[] {
  const out: string[] = []
  let lines: string[]
  try {
    lines = readFileSync(join(base, `${id}.jsonl`), 'utf-8').split('\n')
  } catch {
    return out
  }
  for (const line of lines) {
    if (!line.trim()) continue
    let obj: any
    try { obj = JSON.parse(line) } catch { continue }
    if (obj.type !== 'subagent/spawn') continue
    const childId = obj.data?.session_id
    if (!childId || seen.has(childId)) continue
    seen.add(childId)
    out.push(childId)
    out.push(...subagentSessionIds(childId, base, seen))
  }
  return out
}
